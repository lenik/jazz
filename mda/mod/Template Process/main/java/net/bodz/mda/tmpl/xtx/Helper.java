package net.bodz.mda.tmpl.xtx;

import java.io.IOException;
import java.io.InputStream;

import net.bodz.bas.lang.err.CompileException;
import net.bodz.bas.sys.Processes;
import net.bodz.bas.types.util.Arrays2;

public class Helper {

    static class InterpretedCompiler implements TemplateScript {

        private XtxLang  lang;
        private String[] shellCommand;

        public InterpretedCompiler(XtxLang lang, String[] shellCommand) {
            if (lang == null)
                throw new NullPointerException("lang");
            if (shellCommand == null)
                throw new NullPointerException("shellCommand");
            this.lang = lang;
            this.shellCommand = shellCommand;
        }

        public XtxLang getLanguage() {
            return lang;
        }

        @Override
        public InputStream execute(String... args) throws CompileException, TemplateException,
                IOException {
            if (args == null)
                args = new String[0];
            String[] cmdarray;
            if (args == null)
                cmdarray = args;
            else
                cmdarray = Arrays2.concat(shellCommand, args);
            Process process = Processes.shellExec(cmdarray);
            InputStream in = process.getInputStream();
            return in;
        }
    }

    public static InterpretedCompiler interpretedCompiler(XtxLang lang, String... shellCommand) {
        return new InterpretedCompiler(lang, shellCommand);
    }

}
