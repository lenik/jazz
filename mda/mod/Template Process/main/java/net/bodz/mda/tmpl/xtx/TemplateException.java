package net.bodz.mda.tmpl.xtx;

public class TemplateException extends Exception {

    private static final long serialVersionUID = -749055326331229070L;

    public TemplateException() {
        super();
    }

    public TemplateException(String message, Throwable cause) {
        super(message, cause);
    }

    public TemplateException(String message) {
        super(message);
    }

    public TemplateException(Throwable cause) {
        super(cause);
    }

}
