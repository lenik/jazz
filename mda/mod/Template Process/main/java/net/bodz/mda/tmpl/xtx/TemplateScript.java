package net.bodz.mda.tmpl.xtx;

import java.io.IOException;
import java.io.InputStream;

import net.bodz.bas.lang.err.CompileException;

public interface TemplateScript {

    XtxLang getLanguage();

    InputStream execute(String... args) throws CompileException, TemplateException, IOException;

}
