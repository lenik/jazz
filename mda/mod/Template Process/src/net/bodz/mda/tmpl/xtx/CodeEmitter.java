package net.bodz.mda.tmpl.xtx;

import java.io.IOException;

public interface CodeEmitter {

    void setSourceLocation(String source);

    void setSourcePosition(int line, int column);

    void start();

    void emitCode(String code) throws IOException;

    void emitExpr(String exprCode) throws IOException;

    void emitText(String text) throws IOException;

    void end();

}
