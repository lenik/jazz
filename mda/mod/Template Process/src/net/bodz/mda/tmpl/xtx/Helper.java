package net.bodz.mda.tmpl.xtx;

import net.bodz.bas.lang.ControlExit;
import net.bodz.bas.lang.TRunnable;
import net.bodz.bas.sys.Processes;
import net.bodz.bas.types.util.Arrays2;

public class Helper {

    static class InterpretedCompiler implements TRunnable<String[], Exception> {

        private boolean  waitFor;
        private String[] shellCommand;

        public InterpretedCompiler(boolean waitFor, String[] shellCommand) {
            if (shellCommand == null)
                throw new NullPointerException("shellCommand");
            this.waitFor = waitFor;
            this.shellCommand = shellCommand;
        }

        @Override
        public void run(String[] parameters) throws Exception, ControlExit {
            if (parameters == null)
                parameters = new String[0];
            String[] cmdarray;
            if (parameters == null)
                cmdarray = parameters;
            else
                cmdarray = Arrays2.concat(shellCommand, parameters);
            Process process = Processes.shellExec(cmdarray);
            if (waitFor) {
                int exitValue = process.waitFor();
                throw new ControlExit(exitValue);
            }
        }
    }

    /**
     * @return the returned TRunnable will throw {@link ControlExit} with-in its
     *         {@link #run(String[])} method.
     */
    public static InterpretedCompiler interpretedCompiler(boolean waitFor, String... shellCommand) {
        return new InterpretedCompiler(waitFor, shellCommand);
    }

}
