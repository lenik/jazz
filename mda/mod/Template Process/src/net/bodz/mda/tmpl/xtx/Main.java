package net.bodz.mda.tmpl.xtx;

import java.io.File;
import java.io.IOException;
import java.io.Reader;
import java.io.Writer;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;

import net.bodz.bas.a.Doc;
import net.bodz.bas.a.ProgramName;
import net.bodz.bas.a.RcsKeywords;
import net.bodz.bas.a.Version;
import net.bodz.bas.cli.BasicCLI;
import net.bodz.bas.cli.a.Option;
import net.bodz.bas.io.Files;
import net.bodz.bas.io.CharOuts.BCharOut;
import net.bodz.bas.io.CharOuts.WriterCharOut;
import net.bodz.bas.lang.ControlExit;
import net.bodz.bas.lang.TRunnable;
import net.bodz.bas.lang.err.IllegalUsageError;
import net.bodz.bas.lang.err.IllegalUsageException;
import net.bodz.bas.lang.err.ParseException;
import net.bodz.bas.types.sg.InvalidStateException;
import net.bodz.mda.tmpl.xtx.langs.Perl;
import net.bodz.mda.tmpl.xtx.langs.Python;

@Doc("XTX Interpreter")
@ProgramName("xtx")
@RcsKeywords(id = "$Id$")
@Version( { 0, 1 })
public class Main extends BasicCLI {

    public static final int CODE_GATE = '\u00AC';                // Â¬
    public static final int ESCAPE    = '\\';

    @Option(alias = "e", vnam = "CHARSET", doc = "encoding of template file")
    Charset                 encoding  = Charset.defaultCharset();

    @Option(alias = "c", vnam = "DIR", doc = "where to put compiled templates")
    File                    compileDirectory;

    @Option(alias = "d", vnam = "DIR", doc = "where to put result files, or directory")
    File                    output;

    @Option(vnam = "FILE", fileIndex = 0, required = true)
    File                    template;

    String                  templateNameWithoutXtx;
    File                    compiledFile;
    File                    resultFile;

    @Override
    protected void _boot() throws Exception {
        if (template == null)
            throw new IllegalUsageException("template isn't specified");

        String name = template.getName();
        if (name.endsWith(".xtx"))
            templateNameWithoutXtx = name.substring(0, name.length() - 4);
        else if (name.endsWith("t"))
            templateNameWithoutXtx = name.substring(0, name.length() - 1);
        else
            throw new IllegalUsageError("illegal xtx-template extension");

        if (compileDirectory == null)
            compileDirectory = template.getParentFile();
        compiledFile = new File(compileDirectory, templateNameWithoutXtx);

        if (output == null)
            output = compileDirectory;
        if (!output.isFile()) {
            String resultName = Files.getName(templateNameWithoutXtx);
            output = new File(output, resultName);
        }
    }

    @Override
    protected void doMainManaged(String[] args) throws Exception {
        String name = template.getName();
        XtxLang matchingLang = null;
        for (XtxLang lang : languages) {
            if (lang.matches(name)) {
                matchingLang = lang;
                break;
            }
        }
        if (matchingLang == null)
            throw new UnsupportedOperationException("No matching host lang for name: " + name);

        if (!compiledFile.exists() || //
                template.lastModified() > compiledFile.lastModified()) {
            Reader templateReader = Files.getReader(template, encoding);
            Writer compiledWriter = Files.getWriter(compiledFile, encoding);
            WriterCharOut cout = new WriterCharOut(compiledWriter);
            try {
                CodeEmitter emitter = matchingLang.newEmitter(cout); // options...
                parse(templateReader, emitter);
            } finally {
                templateReader.close();
                cout.close();
            }
        }

        TRunnable<String[], Exception> runnable = matchingLang.compile(compiledFile.getPath());
        if (runnable != null)
            try {
                runnable.run(args);
            } catch (ControlExit exit) {
                int status = exit.getStatus();
                System.exit(status);
            }
    }

    static final int S_TEXT = 0;
    static final int S_CODE = 1;
    static final int S_EXPR = 2;

    public static void parse(Reader reader, CodeEmitter emitter) throws IOException, ParseException {
        if (reader == null)
            throw new NullPointerException("reader");
        int state = S_TEXT;
        int braceLevel = 0;
        BCharOut buffer = new BCharOut();
        int c;
        while ((c = reader.read()) != -1) {
            if (c == ESCAPE) {
                c = reader.read();
                switch (c) {
                case CODE_GATE:
                case '{':
                case '}':
                    buffer._write(c);
                    continue;
                }
                buffer._write(ESCAPE);
            }
            switch (state) {
            case S_TEXT:
                switch (c) {
                case CODE_GATE:
                    state = S_CODE;
                    break;
                case '{':
                    state = S_EXPR;
                    braceLevel++;
                    break;
                default:
                    buffer._write(c);
                    continue;
                }
                String text = buffer.flip();
                emitter.emitText(text);
                break;
            case S_CODE:
                switch (c) {
                case '\n':
                    buffer._write('\n');
                case CODE_GATE:
                    state = S_TEXT;
                    break;
                default:
                    buffer._write(c);
                    continue;
                }
                String code = buffer.flip();
                emitter.emitCode(code);
                break;
            case S_EXPR:
                switch (c) {
                case '{':
                    braceLevel++;
                    break;
                case '}':
                    braceLevel--;
                    break;
                }
                buffer._write(c);
                if (braceLevel == 0) {
                    String exprCode = buffer.flip();
                    emitter.emitExpr(exprCode);
                    state = S_TEXT;
                }
                break;
            default:
                throw new InvalidStateException();
            }
        }
        switch (state) {
        case S_TEXT:
            String text = buffer.flip();
            emitter.emitText(text);
            break;
        case S_CODE:
            String code = buffer.flip();
            emitter.emitCode(code);
            break;
        case S_EXPR:
            throw new ParseException("expect '}': " + buffer);
        }

    }

    public static void main(String[] args) throws Exception {
        new Main().run(args);
    }

    static final List<XtxLang> languages;
    static {
        languages = new ArrayList<XtxLang>();
        languages.add(new Perl());
        languages.add(new Python());
    }

}
