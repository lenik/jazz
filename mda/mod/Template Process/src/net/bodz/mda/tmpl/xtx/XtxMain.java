package net.bodz.mda.tmpl.xtx;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Reader;
import java.io.Writer;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;

import net.bodz.bas.a.Doc;
import net.bodz.bas.a.ProgramName;
import net.bodz.bas.a.RcsKeywords;
import net.bodz.bas.a.Version;
import net.bodz.bas.cli.BasicCLI;
import net.bodz.bas.cli.a.Option;
import net.bodz.bas.io.CharOut;
import net.bodz.bas.io.CharOuts;
import net.bodz.bas.io.Files;
import net.bodz.bas.io.CharOuts.BCharOut;
import net.bodz.bas.io.CharOuts.WriterCharOut;
import net.bodz.bas.lang.err.IllegalUsageError;
import net.bodz.bas.lang.err.IllegalUsageException;
import net.bodz.bas.lang.err.ParseException;
import net.bodz.bas.types.sg.InvalidStateException;
import net.bodz.bas.types.util.Strings;
import net.bodz.mda.tmpl.xtx.langs.Perl;
import net.bodz.mda.tmpl.xtx.langs.Python;

@Doc("XTX Interpreter")
@ProgramName("xtx")
@RcsKeywords(id = "$Id$")
@Version( { 0, 1 })
public class XtxMain extends BasicCLI {

    public static final int CODE_GATE = '\u00AC';                // Â¬
    public static final int ESCAPE    = '\\';

    @Option(alias = "e", vnam = "CHARSET", doc = "encoding of template file")
    Charset                 encoding  = Charset.forName("utf-8");

    @Option(alias = "c", vnam = "DIR", doc = "where to put compiled templates")
    File                    compileDirectory;

    @Option(alias = "t", doc = "only compile the template")
    boolean                 compileOnly;

    @Option(alias = "o", vnam = "DIR", doc = "where to put result files, or directory")
    File                    output;

    @Option(alias = "B", doc = "don't compare with last modified time")
    boolean                 alwaysMake;

    @Option(alias = "s", doc = "send output to stdout")
    boolean                 stdout;

    @Option(vnam = "FILE", fileIndex = 0, required = true)
    File                    template;

    String                  templateNameWithoutXtx;
    File                    scriptFile;
    File                    resultFile;

    @Override
    protected void _boot() throws Exception {
        if (template == null)
            throw new IllegalUsageException("template isn't specified");

        String name = template.getName();
        if (name.endsWith(".xtx"))
            templateNameWithoutXtx = name.substring(0, name.length() - 4);
        else if (name.endsWith("t"))
            templateNameWithoutXtx = name.substring(0, name.length() - 1);
        else
            throw new IllegalUsageError("illegal xtx-template extension");

        if (compileDirectory == null)
            if (output != null && output.isDirectory())
                compileDirectory = output;
            else
                compileDirectory = template.getParentFile();
        scriptFile = new File(compileDirectory, templateNameWithoutXtx);
        L.debug("compile-directory: ", compileDirectory);

        if (output == null)
            output = compileDirectory;
        if (!output.isFile()) {
            String resultName = Files.getName(templateNameWithoutXtx);
            output = new File(output, resultName);
        }
        L.debug("result-output: ", output);
    }

    @Override
    protected void doMain(String[] args) throws Exception {
        L.detail("search matching xtx-language for ", templateNameWithoutXtx);
        XtxLang matchingLang = null;
        for (XtxLang lang : languages) {
            if (lang.matches(templateNameWithoutXtx)) {
                L.detail("found matching xtx-language: ", lang);
                matchingLang = lang;
                break;
            }
        }
        if (matchingLang == null)
            throw new UnsupportedOperationException("No matching host lang for name: "
                    + templateNameWithoutXtx);

        if (alwaysMake || !scriptFile.exists() || //
                template.lastModified() > scriptFile.lastModified()) {
            L.detail("preprocess ", template, " -> ", scriptFile);
            Reader templateReader = Files.getReader(template, encoding);
            CharOut scriptOut;
            if (stdout && compileOnly)
                scriptOut = CharOuts.stdout;
            else {
                Writer scriptWriter = Files.getWriter(scriptFile, encoding);
                scriptOut = new WriterCharOut(scriptWriter);
            }
            try {
                CodeEmitter emitter = matchingLang.newEmitter(scriptOut); // options...
                parse(templateReader, emitter);
            } finally {
                templateReader.close();
                if (scriptOut != CharOuts.stdout)
                    scriptOut.close();
            }
        }
        L.detail("preprocess finished. ");
        if (compileOnly)
            return;

        L.detail("compile using local language ", matchingLang.getName());
        TemplateScript script = matchingLang.compile(scriptFile.getPath());

        L.detail("execute the compiled template");
        if (L.showDebug()) {
            L.debug("    with arguments: ");
            for (int i = 0; i < args.length; i++)
                L.debug("    ", i, ". \"", Strings.escape(args[i]), "\"");
        }
        InputStream resultStream = script.execute(args);
        OutputStream out = null;
        try {
            out = stdout ? System.out : new FileOutputStream(output);
            byte[] block = new byte[4096];
            int cb;
            while ((cb = resultStream.read(block)) != -1) {
                out.write(block, 0, cb);
            }
        } finally {
            resultStream.close();
            if (out != null && out != System.out)
                out.close();
        }
        L.detail("done");
    }

    static final int S_LEAD_TEXT = 0;
    static final int S_TEXT      = 1;
    static final int S_CODE      = 3;
    static final int S_EXPR      = 5;

    public static void parse(Reader reader, CodeEmitter emitter) throws IOException, ParseException {
        if (reader == null)
            throw new NullPointerException("reader");
        int state = S_TEXT;
        int braceLevel = 0;
        BCharOut buffer = new BCharOut();
        BCharOut leadingBuffer = new BCharOut();
        int y = 0;
        int x = 0;
        boolean atLeading = true;
        int c;
        while ((c = reader.read()) != -1) {
            if (c == '\n') {
                y++;
                x = 0;
            } else
                x++;
            if (c == ESCAPE) {
                c = reader.read();
                switch (c) {
                case CODE_GATE:
                case '{':
                case '}':
                    buffer._write(c);
                    continue;
                }
                buffer._write(ESCAPE);
            }
            switch (state) {
            case S_LEAD_TEXT:
                atLeading = x == 0;
                state = S_TEXT;
            case S_TEXT:
                if (atLeading)
                    if (Character.isSpaceChar(c)) {
                        leadingBuffer._write(c);
                        continue;
                    }
                atLeading = false;
                switch (c) {
                case CODE_GATE:
                    state = S_CODE;
                    leadingBuffer.flip(); // skip leading space to the gate.
                    break;
                case '{':
                    state = S_EXPR;
                    braceLevel++;
                    break;
                default:
                    buffer._write(c);
                    continue;
                }
                String text = buffer.flip();
                if (!text.isEmpty())
                    emitter.emitText(text);
                break;
            case S_CODE:
                switch (c) {
                case '\n':
                    buffer._write('\n');
                case CODE_GATE:
                    state = S_LEAD_TEXT;
                    break;
                default:
                    buffer._write(c);
                    continue;
                }
                String code = buffer.flip();
                if (!code.isEmpty())
                    emitter.emitCode(code);
                break;
            case S_EXPR:
                switch (c) {
                case '{':
                    braceLevel++;
                    break;
                case '}':
                    braceLevel--;
                    break;
                }
                if (braceLevel == 0) {
                    String exprCode = buffer.flip();
                    if (!exprCode.isEmpty())
                        emitter.emitExpr(exprCode);
                    state = S_TEXT;
                } else
                    buffer._write(c);
                break;
            default:
                throw new InvalidStateException();
            }
        }
        switch (state) {
        case S_LEAD_TEXT:
        case S_TEXT:
            String text = buffer.flip();
            if (!text.isEmpty())
                emitter.emitText(text);
            break;
        case S_CODE:
            String code = buffer.flip();
            if (!code.isEmpty())
                emitter.emitCode(code);
            break;
        case S_EXPR:
            throw new ParseException("expect '}': " + buffer);
        }
    }

    public static void main(String[] args) throws Exception {
        new XtxMain().run(args);
    }

    static final List<XtxLang> languages;
    static {
        languages = new ArrayList<XtxLang>();
        languages.add(new Perl());
        languages.add(new Python());
    }

}
