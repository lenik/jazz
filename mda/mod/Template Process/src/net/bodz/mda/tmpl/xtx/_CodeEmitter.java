package net.bodz.mda.tmpl.xtx;

import java.io.IOException;
import java.util.Date;

import net.bodz.bas.io.CharOut;
import net.bodz.bas.types.util.Strings;

public abstract class _CodeEmitter implements CodeEmitter {

    private boolean         multiline;
    private int             heredocMinSize = 100;

    private String          source         = "* unknown *";
    private int             line;
    private int             column;

    protected final CharOut out;

    public _CodeEmitter(CharOut out) {
        if (out == null)
            throw new NullPointerException("out");
        this.out = out;
    }

    @Override
    public void setSourceLocation(String source) {
        this.source = source;
    }

    @Override
    public void setSourcePosition(int line, int column) {
        this.line = line;
        this.column = column;
    }

    protected String getSource() {
        return source;
    }

    protected int getLine() {
        return line;
    }

    protected int getColumn() {
        return column;
    }

    @Override
    public void start() {
        emitCommentLine("");
        emitCommentLine("Emitted by: " + getClass().getName());
        emitCommentLine("Source: " + getSource());
        emitCommentLine("");
    }

    @Override
    public void end() {
        emitCommentLine("");
        emitCommentLine("Emit date: " + new Date());
        emitCommentLine("");
    }

    protected abstract String emitCommentLine(String s);

    @Override
    public void emitCode(String code) throws IOException {
        out.print(code);
    }

    @Override
    public final void emitExpr(String exprCode) throws IOException {
        emitPrintExpr(exprCode, false);
    }

    @Override
    public final void emitText(String text) throws IOException {
        if (multiline) {
            int i;
            while ((i = text.indexOf('\n')) != -1) {
                String line = text.substring(0, i);
                text = text.substring(i + 1);
                emitPrintExpr(quote(line), true);
            }
        } else if (text.length() >= heredocMinSize) {
            String heredoc = quoteHeredoc(text);
            if (heredoc != null) {
                emitHereDoc(heredoc);
                return;
            }
        }
        emitPrintExpr(text, false);
    }

    public void emitHereDoc(String doc) throws IOException {
        throw new UnsupportedOperationException("heredoc isn't supported. ");
    }

    public abstract void emitPrintExpr(String exprCode, boolean newLine) throws IOException;

    protected String quote(String s) {
        String escaped = Strings.escape(s);
        String quoted = '"' + escaped + '"';
        return quoted;
    }

    /**
     * @return <code>null</code> if heredoc isn't supported, or the given string can't be converted
     *         to heredoc.
     */
    protected String quoteHeredoc(String s) {
        return s;
    }

}
