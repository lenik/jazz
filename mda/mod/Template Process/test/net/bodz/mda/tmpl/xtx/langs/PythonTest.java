package net.bodz.mda.tmpl.xtx.langs;

import java.io.File;
import java.net.URL;

import net.bodz.bas.io.Files;
import net.bodz.mda.tmpl.xtx.XtxMain;

import org.junit.Test;

public class PythonTest {

    String tmpdir = Files.getTmpDir().getPath();

    @Test
    public void test() throws Exception {
        String encoding = "utf-8";
        URL _template = Files.classData(getClass(), "pyt");
        File template = new File(_template.toURI());

        // String opts = "-B";
        String opts = "-Bst";
        XtxMain.main(new String[] { opts, "-e", encoding, //
                "-o", tmpdir, //
                template.getPath(), });
        File scriptFile = new File(tmpdir, "PythonTest.py");
        File resultFile = new File(tmpdir, "PythonTest");

        String scriptExpected = Files.readAll(Files.classData(getClass(), "py.expected"), encoding);
        String scriptActual = Files.readAll(scriptFile, encoding);
        Files.equals(scriptExpected, scriptActual);

        String resultExpected = Files.readAll(Files.classData(getClass(), "expected"), encoding);
        String resultActual = Files.readAll(resultFile, encoding);
        Files.equals(resultExpected, resultActual);
    }

}
