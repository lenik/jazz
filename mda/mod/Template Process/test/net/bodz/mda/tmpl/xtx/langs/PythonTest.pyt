¬if 1: 
¬    pass
1ok.
¬if 2: 
¬   pass
¬       
2ok.
