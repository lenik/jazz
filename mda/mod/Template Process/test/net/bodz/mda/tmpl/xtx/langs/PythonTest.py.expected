from sys import stdout
if 1: 
    pass
stdout.write("1ok.\n".__str__())
if 2: 
   pass
       
stdout.write("2ok.\n".__str__())
